[JSC] Introduce fast iteration for JSNodeList in "for..of"

hyjorc1 opened the pull request:
#### 37c5a6a2054d85cad14f8f29e4d0eb5091282bc8
<pre>
[JSC] Introduce fast iteration for JSNodeList in &quot;for..of&quot;
<a href="https://bugs.webkit.org/show_bug.cgi?id=262911">https://bugs.webkit.org/show_bug.cgi?id=262911</a>
rdar://116694150

Reviewed by NOBODY (OOPS!).

Since WebCore::NodeList supports length() and item(unsigned index)
for both static and live node list, we should introduce the fast path
of `for..of` iteration for JSNodeList. That means instead of emitting
and using the bytecodes of next() in op_iterator_open and op_iterator_next,
the generic path, we should directly access the next value via length() and
item(unsigned index) during the iteration.

This patch is focusing on two tasks:
1. Introduce JSFastIterable and let JSNodeList and JSRadioNodeList to inherit
   it directly.
2. Split arrayIteratorProtocolWatchpointSet to arrayIteratorNextWatchpointSet
   and arrayIteratorWatchpointSet for more specific purposes.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/bytecode/IterationModeMetadata.h:
(JSC::isFastIteration):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::parseBlock):
* Source/JavaScriptCore/runtime/ClassInfo.h:
* Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
(JSC::iteratorOpenTryFastImpl):
(JSC::iteratorNextTryFastImpl):
* Source/JavaScriptCore/runtime/IteratorOperations.cpp:
(JSC::getFastIterationMode):
(JSC::getIterationMode):
* Source/JavaScriptCore/runtime/IteratorOperations.h:
* Source/JavaScriptCore/runtime/JSCast.h:
* Source/JavaScriptCore/runtime/JSIterableFast.cpp: Added.
(JSC::JSIterableFast::length):
(JSC::JSIterableFast::getIndex):
* Source/JavaScriptCore/runtime/JSIterableFast.h: Added.
(JSC::JSIterableFast::JSIterableFast):
(JSC::JSIterableFast::finishCreation):
(JSC::JSWrapper::wrapped const):
(JSC::JSWrapper::offsetOfWrapped):
(JSC::JSWrapper::hasCustomPtrTraits):
(JSC::JSWrapper::JSWrapper):
(JSC::JSScriptWrappable::offsetOfWrapper):
(JSC::asIterableFast):
(JSC::isIterableFast):
(JSC::canIterateFast):
* Source/JavaScriptCore/runtime/JSIterableFastInlines.h: Added.
(JSC::JSScriptWrappable::wrapper const):
(JSC::JSScriptWrappable::setWrapper):
(JSC::JSScriptWrappable::clearWrapper):
* Source/JavaScriptCore/runtime/JSType.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/runtime/JSType.h:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/WebCore/bindings/js/JSDOMWrapperCache.h:
(WebCore::setInlineCachedWrapper):
(WebCore::clearInlineCachedWrapper):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateHeader):
(GenerateImplementation):
* Source/WebCore/dom/NodeList.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/37c5a6a2054d85cad14f8f29e4d0eb5091282bc8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22200 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22406 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23272 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24097 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20555 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22448 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26661 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22728 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21573 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22433 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/22146 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19261 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24948 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19204 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20117 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26377 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/19307 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20214 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20344 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24238 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/21574 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20785 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17686 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/25617 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20135 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6027 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24346 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/26892 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20733 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5858 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

shvaikalesh commented on Source/WebCore/dom/NodeList.h:
```diff
@@ -41,6 +41,7 @@ class NodeList : public ScriptWrappable, public RefCounted<NodeList> {
     // DOM methods & attributes for NodeList
     virtual unsigned length() const = 0;
     virtual Node* item(unsigned index) const = 0;
+    virtual bool isStaticNode() const { return false; };
```
```suggestion
    virtual bool isStaticNodeList() const { return false; };
```

would be nice to co-locate with `isLiveNodeList` and friends.

hyjorc1 replied:
Maybe we don't need that. Removed.

shvaikalesh commented on Source/WebCore/dom/NodeList.idl:
```diff
@@ -29,5 +29,6 @@
 ] interface NodeList {
     getter Node? item(unsigned long index);
     readonly attribute unsigned long length;
+    readonly attribute boolean isStaticNode;
```
This is observable, and we aren't required to expose every method to JS userland code.

shvaikalesh commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5018,6 +5029,29 @@ sub GenerateImplementation
         push(@implContent, "}\n\n");
     }
 
+    if ($interfaceName eq "NodeList") {
```
Would be nice to move this to `JSNodeListCustom.cpp`.

hyjorc1 replied:
Hmm.. moved to `JSNodeListCustom.cpp`. But got compile error shown in https://ews-build.webkit.org/#/builders/2/builds/22559. Any idea how to resolve that?

Constellation replied:
I think this is simply some headers are missing. Let's see the error message and add missing headers.

shvaikalesh reviewed (commented):
🐈

hyjorc1 committed:
[JSC] Introduce fast iteration for JSNodeList in "for..of"
https://bugs.webkit.org/show_bug.cgi?id=262911
rdar://116694150

Reviewed by NOBODY (OOPS!).

Since WebCore::NodeList supports length() and item(unsigned index)
for both static and live node list, we should introduce the fast path
of `for..of` iteration for JSNodeList. That means instead of emitting
and using the bytecodes of next() in op_iterator_open and op_iterator_next,
the generic path, we should directly access the next value via length() and
item(unsigned index) during the iteration.

This patch is focusing on two tasks:
1. Introduce JSFastIterable and let JSNodeList and JSRadioNodeList to inherit
   it directly.
2. Split arrayIteratorProtocolWatchpointSet to arrayIteratorNextWatchpointSet
   and arrayIteratorWatchpointSet for more specific purposes.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/bytecode/IterationModeMetadata.h:
(JSC::isFastIteration):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::parseBlock):
* Source/JavaScriptCore/runtime/ClassInfo.h:
* Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
(JSC::iteratorOpenTryFastImpl):
(JSC::iteratorNextTryFastImpl):
* Source/JavaScriptCore/runtime/IteratorOperations.cpp:
(JSC::getFastIterationMode):
(JSC::getIterationMode):
* Source/JavaScriptCore/runtime/IteratorOperations.h:
* Source/JavaScriptCore/runtime/JSCast.h:
* Source/JavaScriptCore/runtime/JSIterableFast.cpp: Added.
(JSC::JSIterableFast::length):
(JSC::JSIterableFast::getIndex):
* Source/JavaScriptCore/runtime/JSIterableFast.h: Added.
(JSC::JSIterableFast::JSIterableFast):
(JSC::JSIterableFast::finishCreation):
(JSC::JSWrapper::wrapped const):
(JSC::JSWrapper::offsetOfWrapped):
(JSC::JSWrapper::hasCustomPtrTraits):
(JSC::JSWrapper::JSWrapper):
(JSC::JSScriptWrappable::offsetOfWrapper):
(JSC::asIterableFast):
(JSC::isIterableFast):
(JSC::canIterateFast):
* Source/JavaScriptCore/runtime/JSIterableFastInlines.h: Added.
(JSC::JSScriptWrappable::wrapper const):
(JSC::JSScriptWrappable::setWrapper):
(JSC::JSScriptWrappable::clearWrapper):
* Source/JavaScriptCore/runtime/JSType.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/runtime/JSType.h:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/WebCore/bindings/js/JSDOMWrapperCache.h:
(WebCore::setInlineCachedWrapper):
(WebCore::clearInlineCachedWrapper):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateHeader):
(GenerateImplementation):
* Source/WebCore/dom/NodeList.h:

Constellation commented on Source/JavaScriptCore/runtime/ClassInfo.h:
```diff
@@ -173,10 +179,15 @@ struct MethodTable {
         &ClassName::visitOutputConstraints, \
         &ClassName::visitOutputConstraints, \
+        LengthPtr, \
+        GetIndexPtr, \
```
Let's align it with the above. And let's define these functions in JSCell too (with unreachable assertion).
Maybe, we should name them to more explicit ones, like `getFastIterableLength` / `getFastIterableIndexedElement`.

Jarred-Sumner commented on Source/JavaScriptCore/runtime/ClassInfo.h:
```diff
@@ -123,6 +123,12 @@ struct MethodTable {
+    using GetFastIterableLengthPtr = unsigned (*)(JSObject*);
+    GetFastIterableLengthPtr METHOD_TABLE_ENTRY(getFastIterableLength);
+
+    using GetFastIterableIndexedElementPtr = JSValue (*)(JSGlobalObject*, JSObject*, unsigned index);
```
This is very nice
